In [37]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math 

In [13]:
tweetsData=pd.read_csv("data.csv")

In [18]:
# Let's see number of observations(size of data) we are dealing with 
tweetsData.shape

(25694, 34)

So we are having 25694 observations over 34 distinct parameters

In [17]:
# Let see if what are the attributes we are dealing with
tweetsData.columns

Index(['tweet_id', 'tweet_created_at', 'tweet_created_on_holiday_bool',
       'tweet_created_on_weekend_bool', 'tweet_created_at_noon_bool',
       'tweet_created_at_eve_bool', 'user_id', 'user_screen_name',
       'user_screen_name_length', 'user_no_of_tweets', 'user_no_of_followers',
       'user_no_of_followings', 'user_account_age', 'user_no_of_favourites',
       'user_average_tweets', 'user_average_favourites',
       'user_account_location', 'tweet_text', 'tweet_text_length',
       'tweet_text_optimal_length', 'tweet_text_no_of_hashtags',
       'tweet_text_contains_hashtags', 'tweet_text_contains_url',
       'tweet_text_no_of_user_mentions', 'tweet_text_contains_user_mentions',
       'tweet_text_sentiment', 'tweet_text_contains_media',
       'tweet_text_contains_number', 'tweet_text_contains_upper_words',
       'tweet_text_contains_lower_words', 'tweet_text_contains_excl',
       'tweet_text_contains_retweet_suggestion', 'retweeted', 'retweets'],
      dtype='object')

In [15]:
# Let's take a basic look at this columns
tweetsData.head()

tweet_id                tweet_created_at  \
0  ~1357587318119407616~  Fri Feb 05 07:10:14 +0000 2021   
1  ~1357587287681323010~  Fri Feb 05 07:10:06 +0000 2021   
2  ~1357587268689485825~  Fri Feb 05 07:10:02 +0000 2021   
3  ~1357587262469525507~  Fri Feb 05 07:10:00 +0000 2021   
4  ~1357587255699775488~  Fri Feb 05 07:09:59 +0000 2021   

   tweet_created_on_holiday_bool  tweet_created_on_weekend_bool  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   tweet_created_at_noon_bool  tweet_created_at_eve_bool  \
0                           0                          0   
1                           0                          0   
2                           0                          0   
3                           0                          0   
4                           0                          0   

                 user_id user_screen_name  user_screen_name_length  \
0            ~327890669~         mavirise                        8   
1            ~327890669~         mavirise                        8   
2  ~1216545386040418304~     aman89086818                       12   
3           ~1731144182~   jagbirsingh111                       14   
4            ~327890669~         mavirise                        8   

   user_no_of_tweets  ...  tweet_text_contains_user_mentions  \
0               4355  ...                                  0   
1               4355  ...                                  0   
2                 48  ...                                  0   
3                196  ...                                  1   
4               4355  ...                                  0   

   tweet_text_sentiment  tweet_text_contains_media  \
0                   nue                          0   
1                   nue                          0   
2                   nue                          0   
3                   neg                          0   
4                   nue                          0   

   tweet_text_contains_number  tweet_text_contains_upper_words  \
0                           1                                0   
1                           1                                0   
2                           0                                0   
3                           0                                0   
4                           1                                0   

   tweet_text_contains_lower_words tweet_text_contains_excl  \
0                                1                        1   
1                                1                        1   
2                                1                        0   
3                                1                        0   
4                                1                        1   

  tweet_text_contains_retweet_suggestion  retweeted  retweets  
0                                      0      False         0  
1                                      0      False         0  
2                                      0      False         0  
3                                      0      False         0  
4                                      0      False         0  

[5 rows x 34 columns]

In [21]:
# Let us see how many missing values each column have
tweetsData.isna().sum()

tweet_id                                      0
tweet_created_at                              0
tweet_created_on_holiday_bool                 0
tweet_created_on_weekend_bool                 0
tweet_created_at_noon_bool                    0
tweet_created_at_eve_bool                     0
user_id                                       0
user_screen_name                              0
user_screen_name_length                       0
user_no_of_tweets                             0
user_no_of_followers                          0
user_no_of_followings                         0
user_account_age                              0
user_no_of_favourites                         0
user_average_tweets                           0
user_average_favourites                       0
user_account_location                     11436
tweet_text                                    0
tweet_text_length                             0
tweet_text_optimal_length                     0
tweet_text_no_of_hashtags               

In [56]:
len(tweetsData[tweetsData.retweets==0])

15827

So we can see that we only have missing values in user_account_location which is approx 40% which is due to the user not given permission to let twitter know user's account location. However this will not matter a lot here. Also retweets column has no missing value.

In [19]:
# Let's have a look at data types of each parameter and number of non null entries
tweetsData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25694 entries, 0 to 25693
Data columns (total 34 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   tweet_id                                25694 non-null  object 
 1   tweet_created_at                        25694 non-null  object 
 2   tweet_created_on_holiday_bool           25694 non-null  int64  
 3   tweet_created_on_weekend_bool           25694 non-null  int64  
 4   tweet_created_at_noon_bool              25694 non-null  int64  
 5   tweet_created_at_eve_bool               25694 non-null  int64  
 6   user_id                                 25694 non-null  object 
 7   user_screen_name                        25694 non-null  object 
 8   user_screen_name_length                 25694 non-null  int64  
 9   user_no_of_tweets                       25694 non-null  int64  
 10  user_no_of_followers                    25694 non-null  in

In [20]:
# Let us have a brief overview of the summary statistics over NUMERIC columns in twitter data
tweetsData.describe()

tweet_created_on_holiday_bool  tweet_created_on_weekend_bool  \
count                        25694.0                   25694.000000   
mean                             0.0                       0.015256   
std                              0.0                       0.122574   
min                              0.0                       0.000000   
25%                              0.0                       0.000000   
50%                              0.0                       0.000000   
75%                              0.0                       0.000000   
max                              0.0                       1.000000   

       tweet_created_at_noon_bool  tweet_created_at_eve_bool  \
count                25694.000000               25694.000000   
mean                     0.096404                   0.080213   
std                      0.295150                   0.271628   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      1.000000                   1.000000   

       user_screen_name_length  user_no_of_tweets  user_no_of_followers  \
count             25694.000000       25694.000000          2.569400e+04   
mean                 11.995135        5521.590255          2.410177e+03   
std                   2.567414       20504.915268          8.179765e+04   
min                   4.000000           1.000000          0.000000e+00   
25%                  10.000000          95.000000          8.000000e+00   
50%                  12.000000         651.000000          4.900000e+01   
75%                  14.000000        4019.000000          2.507500e+02   
max                  15.000000      829961.000000          5.483319e+06   

       user_no_of_followings  user_account_age  user_no_of_favourites  ...  \
count           25694.000000      25694.000000           25694.000000  ...   
mean              402.654705       1270.251382            5792.574103  ...   
std              1219.261763       1406.534540           18283.982312  ...   
min                 0.000000          0.000000               0.000000  ...   
25%                38.000000         67.000000              97.000000  ...   
50%               118.000000        487.000000             666.000000  ...   
75%               345.750000       2391.750000            3854.500000  ...   
max             79636.000000       5044.000000          451377.000000  ...   

       tweet_text_contains_url  tweet_text_no_of_user_mentions  \
count             25694.000000                    25694.000000   
mean                  0.619717                        0.657975   
std                   0.485466                        1.153367   
min                   0.000000                        0.000000   
25%                   0.000000                        0.000000   
50%                   1.000000                        0.000000   
75%                   1.000000                        1.000000   
max                   1.000000                       11.000000   

       tweet_text_contains_user_mentions  tweet_text_contains_media  \
count                       25694.000000               25694.000000   
mean                            0.385615                   0.119327   
std                             0.486750                   0.324180   
min                             0.000000                   0.000000   
25%                             0.000000                   0.000000   
50%                             0.000000                   0.000000   
75%                             1.000000                   0.000000   
max                             1.000000                   1.000000   

       tweet_text_contains_number  tweet_text_contains_upper_words  \
count                25694.000000                     25694.000000   
mean                     0.663813   

In [50]:
bins=[0,1000,5000,10000,100000]
labs = ["Ordinary","Getting Attention","Popular","Trending"]

In [51]:
tweetsData["Attention_gathered"]=pd.cut(tweetsData.retweets,bins,labels=labs)
tweetsData.Attention_gathered.unique()

[NaN, Ordinary, Getting Attention, Popular, Trending]
Categories (4, object): [Ordinary < Getting Attention < Popular < Trending]

In [52]:
tweetsData['Attention_gathered'].unique()

[NaN, Ordinary, Getting Attention, Popular, Trending]
Categories (4, object): [Ordinary < Getting Attention < Popular < Trending]

In [53]:
df=dict(tuple(tweetsData.groupby('Attention_gathered')))

In [59]:
# Seems Greta is quite getting attention of multiple people. 
df['Trending']

tweet_id                tweet_created_at  \
7444   ~1357282507616645122~  Thu Feb 04 10:59:01 +0000 2021   
10095  ~1357054451769606147~  Wed Feb 03 19:52:48 +0000 2021   

       tweet_created_on_holiday_bool  tweet_created_on_weekend_bool  \
7444                               0                              0   
10095                              0                              0   

       tweet_created_at_noon_bool  tweet_created_at_eve_bool  \
7444                            0                          0   
10095                           0                          1   

                     user_id user_screen_name  user_screen_name_length  \
7444   ~1006419421244678144~    GretaThunberg                       13   
10095  ~1006419421244678144~    GretaThunberg                       13   

       user_no_of_tweets  ...  tweet_text_sentiment  \
7444                8876  ...                   neg   
10095               8876  ...                   neg   

       tweet_text_contains_media  tweet_text_contains_number  \
7444                           0                           1   
10095                          0                           1   

       tweet_text_contains_upper_words  tweet_text_contains_lower_words  \
7444                                 1                                1   
10095                                0                                1   

       tweet_text_contains_excl tweet_text_contains_retweet_suggestion  \
7444                          0                                      1   
10095                         0                                      0   

      retweeted  retweets  Attention_gathered  
7444      False     62666            Trending  
10095     False     13986            Trending  

[2 rows x 35 columns]

Since Greta is getting lot of attention. Let's see what she her hashtags 

In [60]:
df['Trending'].tweet_text_sentiment

7444     neg
10095    neg
Name: tweet_text_sentiment, dtype: object

In [61]:
df['Trending'].tweet_text

7444     I still #StandWithFarmers and support their pe...
10095    Here’s an updated toolkit by people on the gro...
Name: tweet_text, dtype: object

So her tweets are regarding the Farmers Protest going on in India

In [63]:
df['Trending'].tweet_text_contains_url

7444     1
10095    1
Name: tweet_text_contains_url, dtype: int64

In [34]:
df["Popular"]

tweet_id                tweet_created_at  \
1757   ~1357531309745209349~  Fri Feb 05 03:27:40 +0000 2021   
4282   ~1357373229044625409~  Thu Feb 04 16:59:31 +0000 2021   
6448   ~1357301622637006849~  Thu Feb 04 12:14:58 +0000 2021   
10914  ~1356997985662558213~  Wed Feb 03 16:08:26 +0000 2021   

       tweet_created_on_holiday_bool  tweet_created_on_weekend_bool  \
1757                               0                              0   
4282                               0                              0   
6448                               0                              0   
10914                              0                              0   

       tweet_created_at_noon_bool  tweet_created_at_eve_bool  \
1757                            0                          0   
4282                            0                          0   
6448                            1                          0   
10914                           0                          0   

                     user_id user_screen_name  user_screen_name_length  \
1757            ~2314959096~    boxervijender                       13   
4282             ~574795929~      meenaharris                       11   
6448   ~1134059457191776257~     vanessa_vash                       12   
10914   ~810585618031984640~   Jamie_Margolin                       14   

       user_no_of_tweets  ...  tweet_text_sentiment  \
1757                7528  ...                   neg   
4282                9754  ...                   neg   
6448               12799  ...                   neg   
10914              21112  ...                   nue   

       tweet_text_contains_media  tweet_text_contains_number  \
1757                           0                           0   
4282                           0                           1   
6448                           0                           0   
10914                          0                           1   

       tweet_text_contains_upper_words  tweet_text_contains_lower_words  \
1757                                 0                                1   
4282                                 1                                1   
6448                                 1                                1   
10914                                0                                1   

       tweet_text_contains_excl tweet_text_contains_retweet_suggestion  \
1757                          0                                      0   
4282                          0                                      1   
6448                          0                                      1   
10914                         1                                      0   

      retweeted  retweets  Attention_gathered  
1757      False      6388             Popular  
4282      False      8280             Popular  
6448      False      5598             Popular  
10914     False      7277             Popular  

[4 rows x 35 columns]

So all the tweets regarding Farmer protest are trending and popular.

In [41]:
tweetsData['Attention_gathered'].value_counts()

Ordinary             9848
Getting Attention      13
Popular                 4
Trending                2
Name: Attention_gathered, dtype: int64

In [57]:
tweetsData['Attention_gathered'].isna().sum()

15827

So basically 15827 tweets were not retweeted. Those accounts may have less followers.